source/Octopus.Cli/Util/RepositoryExtensions.cs

```diff
@@ -0,0 +1,145 @@
+using Octopus.Client.Extensibility;
+using Octopus.Client.Logging;
+using Octopus.Client.Model;
+using Octopus.Client.Repositories.Async;
+
+namespace Octopus.Cli.Util
+{
+    public static class RepositoryExtensions
+    {
+        private static readonly ILog Logger = LogProvider.GetLogger(typeof(RepositoryExtensions));
+
+        private static async Task<TResource> FindByNameOrIdOrFail<T, TResource>(this T repository,
+            Func<string, Task<TResource>> findByNameFunc, string fixedIdPrefix, string typeDescription, string nameOrId,
+            string inDescription = "", bool skipLog = false)
+            where T : IGet<TResource> where TResource : Resource, INamedResource
+        {
+            TResource resourceById;
+            if (!Regex.IsMatch(nameOrId, $@"^{Regex.Escape(fixedIdPrefix)}-\d+$",
+                RegexOptions.Compiled | RegexOptions.IgnoreCase))
+            {
+                resourceById = null;
+            }
+            else
+            {
+                try
+                {
+                    resourceById = await repository.Get(nameOrId)
+                        .ConfigureAwait(false);
+                }
+                catch (OctopusResourceNotFoundException)
+                {
+                    resourceById = null;
+                }
+            }
+
+            var resourceByName = await findByNameFunc(nameOrId)
+                .ConfigureAwait(false);
+
+            if (resourceById == null && resourceByName == null)
+            {
+                throw new CouldNotFindException(typeDescription, nameOrId, inDescription);
+            }
+
+            if (resourceById != null
+                && resourceByName != null
+                && !string.Equals(resourceById.Id, resourceByName.Id, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new CommandException(
+                    $"Ambiguous {typeDescription} reference '{nameOrId}' matches one {typeDescription} by name and another by id.");
+            }
+
+            var found = resourceById ?? resourceByName;
+
+            if (!skipLog)
+            {
+                Logger.Debug($"Found {typeDescription}: {found.Name} ({found.Id})");
+            }
+
+            return found;
+        }
+
+        private static async Task<TResource[]> FindByNamesOrIdsOrFail<T, TResource>(this T repository,
+            Func<string, Task<TResource>> findByNameFunc, string fixedIdPrefix, string typeDescription,
+            IEnumerable<string> namesOrIds, string inDescription = "", bool skipLog = false)
+            where T : IGet<TResource> where TResource : Resource, INamedResource
+        {
+            var results = await Task.WhenAll(namesOrIds.Select<string, Task<(string missing, TResource resource)>>(
+                async nameOrId =>
+                {
+                    try
+                    {
+                        return (null, await repository.FindByNameOrIdOrFail(findByNameFunc, fixedIdPrefix,
```

Comment:
This returning-null-as-the-first-arg isn't very clear. Consider refactoring.
(if we can avoid null's completely, even better)

Reply:
Okay. I've added a class to make it clearer.